feat(lib): Add List item component (#3618)
Orange-OpenSource/Orange-Boosted-Bootstrap: site/src/content/docs/components/list-item.mdx

---

> title: List item

Reviewer: In the end, I think the name of the group should be Items

Yes it is not very descriptive but it's the common thing with these components.

Right now we have list item components that can be outside a list ! This is not understandable.

I understand the name of the design group is List item but I think we should still rename this page to Items without changing the design name (even though I think we should try to convince Maxime)

In any case we should have a plural like we determined for groups pages (like Badges, Buttons...)

Reviewer 2: It's a problem to break the plural rule that we had until now... And yes, the name is hard to understand.

Author: Should be good now

> description: List items are a flexible and powerful component for displaying a series of content. Modify and extend them to support just about any content within.

Reviewer: The thing is, card items do not necessarily display in a list so where is the series of content?

Reviewer: All descriptions should be updated.

aliases:
  - "/docs/components/item/"
toc: true
types:
  - Navigation card item
  - Navigation list item
  - Static card item
  - Static list item
---

import { getVersionedDocsPath } from '@libs/path'

## Overview

### Component types

> <ul class="list-unstyled row row-cols-1 sm:row-cols-2">

Reviewer: Maybe in this case we should opt for a 2x2 grid to show all the component types (Static card item, static list item, navigation card item, navigation list item)
Also maybe for the lists showing two elements would help?

This would help make the users understand the inherent multiple facets of these components

Reviewer: Tried something

        </div>
      </div>
      <div class="card-body border-default border-thin border-top"><a class="stretched-link" href="#navigation-list-item">Navigation item</a> is a UI element that displays a unit of information and allows users to navigate.</div>
    </div>
  </li>
</ul>

For the two types of items, OUDS Web has a common base `.item`, and `.item-main` classes that set up default styles such as padding and content alignment. The `.item` class is intended to be used in conjunction with our item variants, or to serve as a basis for your own custom styles. The most basic item is the `.item-navigation` variant.

<Example showToolbar={false} code={`<div class="item-wrapper">
    <div class="item">
      <div class="item-main">
        <div class="item-text-container">
          <p class="item-label">Label</p>
        </div>
      </div>
    </div>
  </div>`} />

<Example showToolbar={false} code={`<div class="item-wrapper item-navigation">
    <div class="item">
      <div class="item-main">
        <div class="item-text-container">
          <a href="#" class="item-label item-interactive">Label</a>
        </div>
      </div>
    </div>
  </div>`} />

### Approach

> There are 2 main categories inside the item group of components:

Reviewer 3: I'm a bit confused by this section. I don't know what card item refer to, and there is no other mention of it in the whole doc.
Is the card item a component not implemented yet ? If so, we should maybe specify it.

Reviewer: The items are either Navigation (interactive with a chevron) or Static (no interactivity)
they also are either Card (standalone) or List (in a list)

So you can have a Navigation list item or a Static card item for example.

I will try to propose a table for this section as I think it could be a better way to convey this modularity

Reviewer 3: I think it's clearer, I'm closing this thread.

- The [card item](#card-item) is a single item that can be used as a standalone component.
- The [list item](#list-item) is designed to be used within a list, allowing multiple items to be displayed together.

[Static items](#static-list-item) and [Navigation items](#navigation-list-item) can be used in both categories.

Consider [Navigation items](#navigation-list-item) as an extension of [Static items](#static-list-item), with the added functionality of being able to navigate to different pages or sections within a multi-page interface.

The basic structure includes:
* `.item-wrapper` as the outermost container, which is used to add additional styles or behavior.
* `.item` as the container before the divider.
* `.item-main` as the main container for the primary content of the item.
* `.item-text-container` as the container for the text content of the item.
* `.item-interactive` as the class for interactive elements within the item, such as links or buttons. It's mandatory on the link for the [navigation item](#navigation-list-item).

#### Card item

The card item is the default item when the item components are used.

> Its behavior changes once used in a list. Most of the variants can be applied to this kind of item. It comes with a background and a divider by default.

Reviewer: Does its behavior changes or does it becomes a list item? I think we should delete the first sentence

Reviewer: Tried something

## [[comp]] Static list item

This is used to display a unit of information without any interaction. It can be used as a [standalone component](#card-item) or within a [list](#list-item).

Here is the base DOM and classes you should use for a static item.

All the available variants can be found in the [common parts section](#common-parts).

<Example buttonLabel="standard static items display" class="d-flex flex-column gap-medium" code={`<!-- Card -->
  <div class="item-wrapper">
    <div class="item">
      <div class="item-main">
        <div class="item-text-container">
          <p class="item-label">Label</p>
        </div>
      </div>
    </div>
  </div>

  <!-- List -->
  <ul class="item-list">
    <li class="item-wrapper">
      <div class="item">
        <div class="item-main">
          <div class="item-text-container">
            <p class="item-label">Label</p>
          </div>
        </div>
      </div>
    </li>
    <li class="item-wrapper">
      <div class="item">
        <div class="item-main">
          <div class="item-text-container">
            <p class="item-label">Label</p>
          </div>
        </div>
      </div>
    </li>
  </ul>`} />

> ## [[comp]] Navigation list item

Reviewer: For now I think this part should have a "Navigation items" heading (same for Static list item -> "Static items" before)

#### Back chevron

Add `.item-previous` to the `.item-navigation` to indicate that the item is a back navigation item. This will add a back chevron icon to the start of the item.

<Example buttonLabel="navigation item with back chevron" class="d-flex flex-column gap-medium" code={`<div class="item-wrapper item-navigation item-previous">
    <div class="item">
      <div class="item-main">
        <div class="item-text-container">
          <a href="#" class="item-label item-interactive">Label</a>
        </div>
      </div>
    </div>
  </div>`} />

#### External link

Add `.item-external` to the `.item-navigation` to indicate that the item is an external link. This will add an external link icon to the end of the item.

<Example buttonLabel="navigation item with external link" class="d-flex flex-column gap-medium" code={`<div class="item-wrapper item-navigation item-external">
    <div class="item">
      <div class="item-main">
        <div class="item-text-container">
          <a href="#" class="item-label item-interactive">Label</a>
        </div>
      </div>
    </div>
  </div>`} />

### Layout

#### Using `<a>` around

> Optionally, the `<a>` element can wrap the entire `.item` instead of just the label.

Reviewer: Is there a use case where that is better? If not do we really need to document this?

    <a href="#" class="item item-interactive">
      <div class="item-main">
        <div class="item-text-container">
          <p class="item-label">Label</p>
        </div>
      </div>
    </a>
  </div>`} />

## Common parts

These variants can be used in both [Static items](#static-list-item) and [Navigation items](#navigation-list-item) whether they are [Card items](#card-item) or [List items](#list-item).

### Variants

#### With asset

The main container `.item-main` of an item can contain up to 3 elements. They always appear in the following order:
1. [Leading container](#leading-container) using `.item-leading-container` (optional)
2. [Text container](#texts) using `.item-text-container` (required)
3. [Trailing container](#trailing-container) using `.item-trailing-container` (optional)

##### Leading container

The leading container uses `.item-leading-container`.

Here is a list of the different assets that we support in the leading container, along with their corresponding classes and notes.

You can navigate through the different assets using the tabs below.

<Callout type="warning" title="Small list item">

>   Please note that the sizes below shouldn't be used in a [small list item](#sizes). The slot shouldn't be used as well in this size.

Reviewer 3: I think there is a typo, shouldn't it be "the assets below" ?

Author: I really meant the different sizing classes, the rule is the following: 

The size of the assets is only small when the list item is in small size.

Reviewer 3: Then I suggest clarifying this, that's not what I understand from the callout.
```suggestion
  Please note that the sizes below shouldn't be used in a [small list item](#sizes), only small assets may be used with small list items. The slot shouldn't be used as well in this size.
```

          </div>
        </div>
      </div>
    </div>

    <div class="item-wrapper">
      <div class="item item-no-bg">
        <div class="item-main">
          <div class="item-leading-container item-slot">
            <!-- Disabled via its attribute -->
            <button class="btn btn-strong" disabled>Slot</button>
          </div>
          <div class="item-text-container">
            <p class="item-label">Label</p>
            <p class="item-extra-label">Extra label</p>
          </div>
          <div class="item-trailing-container">
            <!-- Disabled via its style -->
            <p class="tag">Tag</p>
          </div>
        </div>
      </div>
    </div>
  </div>`} />

#### Skeleton

<SkeletonRedirect component="item" />

<Example buttonLabel="item skeleton" class="d-flex flex-column gap-medium" code={`<div aria-busy="true" inert>
    <div class="item-wrapper">
      <div class="item">
        <div class="item-main">
          <div class="item-text-container">
            <p class="item-label">Label</p>
          </div>
        </div>
      </div>
    </div>
  </div>`} />

> ### Sizes

Reviewer 3: I know there is a common template to component pages.
However, sizes section is mentioned repeatedly before being read.
Maybe we should invert sizes and variants section to clarify the available sizes ?

Author: Yes maybe, let's discuss it with other people

Reviewer: I'm not sure, in the Sizes paragraph we reference other variant so we would have the same problem but reversed... Although Sizes are always short paragraphs compared to Variants so it could make sense to switch

Reviewer 2: I understand the reasons, but we should keep the same order everywhere. 
I have the intuition that when we develop, we rather choose first the variant and then the size ? Though I have a small preference for keeping the current order.

                    <p class="item-label">Label</p>
                  </div>
                </div>
              </div>
            </li>
          </ul>
        </div>
      </div>
      <div class="card-body border-default border-thin border-top"><a class="stretched-link" href="#navigation-list-item">Navigation list item</a> is a UI element that displays a list of information and allows users to navigate.</div>
    </div>
  </li>
</ul>

For each types of items, OUDS Web has a common base `.item`, and `.item-main` classes that set up default styles such as padding and content alignment. The `.item` class is intended to be used in conjunction with our item variants, or to serve as a basis for your own custom styles. The most basic item is the `.item-navigation` variant.

### Approach

There are 2 main categories inside the items group of components:
- [Card items](#card-items) are single items that can be used as standalone components.
- [List items](#list-items) are designed to be used within a list, allowing multiple items to be displayed together.

Inside these categories, there are 2 sub-categories:
- Navigation items, which allow navigation.

> - Static items, which are not interactive.

Reviewer 2: Should have static items first to respect the presentation order? (and static is simpler than interactive)

The following table summarizes the different types of items and their categories.

<BsTable>
  | Categories     | Card                 | List                 |
  | -------------- | -------------------- | -------------------- |
  | *Static*       | [Static card item](#static-card-item)     | [Static list item](#static-list-item)     |
  | *Navigation*   | [Navigation card item](#navigation-card-item) | [Navigation list item](#navigation-list-item) |
</BsTable>

The basic structure includes:

> * `.item-wrapper` as the outermost container, which is used to add additional styles or behavior.

Reviewer 2: ```suggestion
* `.item-wrapper` as the external container, which is used to add additional styles or behavior.
```
Just a detail, but more easy to read and understand for non english readers

* `.item` as the container before the divider.
* `.item-main` as the main container for the primary content of the item.
* `.item-text-container` as the container for the text content of the item.
* `.item-interactive` as the class for interactive elements within the item, such as links or buttons. It's mandatory on the link for the [navigation item](#navigation-list-item).

### Accessibility

For navigation items, a wrapping `<nav>` element can be added if it is an important navigation in the page.

> In case of a complex navigation list item, you probably should link all the texts and components together by adding `[aria-describedby]` on the link in order to provide a valid accessible name for the link. The `[aria-describedby]` should point to the id of the element that contains the texts of the link.

Reviewer 2: ```suggestion
In case of a complex navigation list item, you probably should link all the texts and components together by adding `[aria-describedby]` on the link in order to provide a valid accessible name for the link. The `[aria-describedby]` should point to the ids of the elements that contain the texts of the link.
```

            <a href="#" class="item-label item-interactive">Label</a>
          </div>
        </div>
      </div>
    </li>
  </ul>`} />

### Divider

You can remove the divider by using `.item-no-divider` on the `.item-wrapper` or `.item-list`.

<Example buttonLabel="items with no divider" class="d-flex flex-column gap-medium" code={`<!-- Card -->
  <div class="item-wrapper item-no-divider">
    <div class="item">
      <div class="item-main">
        <div class="item-text-container">
          <p class="item-label">Label</p>
        </div>
      </div>
    </div>
  </div>

  <!-- List -->
  <ul class="item-list item-no-divider">
    <li class="item-wrapper">
      <div class="item">
        <div class="item-main">
          <div class="item-text-container">
            <p class="item-label">Label</p>
          </div>
        </div>
      </div>
    </li>
    <li class="item-wrapper item-navigation">
      <div class="item">
        <div class="item-main">
          <div class="item-text-container">
            <a href="#" class="item-label item-interactive">Label</a>
          </div>
        </div>
      </div>
    </li>
  </ul>`} />

### Outlined

> <Callout type="warning" title="Card item only" action="Card item" actionHref="#card-items" />

Reviewer 4: Maybe we could be a bit more verbose here :)

By default every item comes with a divider. The divider may be removed or transformed into an outline using `.item-outlined` and/or `.item-no-outline` on the element `.item-wrapper` or any ancestors. `.item-no-divider` can also be applied for another rendering.

Static items only have the non-interactive version.

<Example buttonLabel="items outlined" class="d-flex flex-column gap-medium" code={`<div class="item-wrapper item-navigation item-outlined">
    <div class="item">
      <div class="item-main">
        <div class="item-text-container">
          <a href="#" class="item-label item-interactive">Label</a>
        </div>
      </div>
    </div>
  </div>

>   <div class="item-wrapper item-navigation item-outlined item-no-outline">

Reviewer 4: I remember talking about this case during the component presentation, but I'm not able to put the component in this situation in Figma. How can this be achieved in figma with current properties ?

Author: Placed a TODO in the code to think about letting it or removing it

          </div>
        </div>
      </div>
    </div>

    <div class="item-wrapper">
      <div class="item item-no-bg">
        <div class="item-main">
          <div class="item-leading-container item-slot">
            <!-- Disabled via its attribute -->
            <button class="btn btn-strong" disabled>Slot</button>
          </div>
          <div class="item-text-container">
            <p class="item-label">Label</p>
            <p class="item-extra-label">Extra label</p>
          </div>
          <div class="item-trailing-container">
            <!-- Disabled via its style -->
            <p class="tag">Tag</p>
          </div>
        </div>
      </div>
    </div>
  </div>`} />

#### Skeleton

<SkeletonRedirect component="item" />

<Example buttonLabel="item skeleton" class="d-flex flex-column gap-medium" code={`<div aria-busy="true" inert>
    <div class="item-wrapper">
      <div class="item">
        <div class="item-main">
          <div class="item-text-container">
            <p class="item-label">Label</p>
          </div>
        </div>
      </div>
    </div>
  </div>`} />

### Sizes

> By default every item comes with a default size. The size can be adapted using `.item-small` on the element `.item-wrapper` or `.item-list`.

Reviewer 4: FYI I've added a comment in Figma because there is an error in the *Navigation card item* since its default size should be "default" and it's actually "small" when we add a new instance.

        <div class="item-main">
          <div class="item-text-container">
            <p class="item-label">Label</p>
          </div>
        </div>
      </div>
    </li>
  </ul>`} />

## Navigation items variants

These variants are exclusive to the navigation items. They are used to indicate a specific type of navigation. The other common variants can be found in the [common variants section](#common-variants).

### Back chevron

Add `.item-previous` to the `.item-navigation` to indicate that the item is a back navigation item. This will add a back chevron icon to the start of the item.

<Example buttonLabel="navigation card item with back chevron" class="d-flex flex-column gap-medium" code={`<div class="item-wrapper item-navigation item-previous">
    <div class="item">
      <div class="item-main">
        <div class="item-text-container">
          <a href="#" class="item-label item-interactive">Label</a>
        </div>
      </div>
    </div>
  </div>`} />

### External link

Add `.item-external` to the `.item-navigation` to indicate that the item is an external link. This will add an external link icon to the end of the item.

<Callout type="info">

>   For accessibility reasons, it is recommended to add a visually hidden text (or a visible one if the icon is far from the text) to indicate that the link opens an external resource. Take a look at [some more complete accessibility guidelines](https://a11y-guidelines.orange.com/en/web/develop/common-navigation/#warn-the-user-when-opening-a-new-window).

Reviewer 2: ```suggestion
  For accessibility reasons, it is recommended to add a `.visually-hidden` text (or a visible one if the icon is far from the text) to indicate that the link opens an external resource. Take a look at [some more complete accessibility guidelines](https://a11y-guidelines.orange.com/en/web/develop/common-navigation/#warn-the-user-when-opening-a-new-window).
```

</Callout>

<Example buttonLabel="navigation card item with external link" class="d-flex flex-column gap-medium" code={`<div class="item-wrapper item-navigation item-external">
    <div class="item">
      <div class="item-main">
        <div class="item-text-container">
          <a href="#" target="_blank" rel="noopener" class="item-label item-interactive">Label<span class="visually-hidden">&nbsp;(external link, new window)</span></a>
        </div>
      </div>
    </div>
  </div>

  <div class="item-wrapper item-navigation item-external">
    <div class="item">
      <div class="item-main">
        <div class="item-text-container">
          <a href="#" target="_blank" rel="noopener" class="item-label item-interactive">Label (external link, new window)</a>
        </div>
      </div>
    </div>
  </div>`} />

### Wrapping link

Optionally, the `<a>` element can wrap the entire `.item` instead of just the label.

<Callout type="info" title="Accessibility">

>   ATs will read the entire item as a link. You can use the visually hidden text to provide additional context if needed. You won't need any `[aria-describedby]` as stated in our [accessibility section](#accessibility).

Reviewer 2: ```suggestion
  Screen readers will read the entire item as a link. You can use the `.visually-hidden` text to provide additional context if needed. You won't need any `[aria-describedby]` as stated in our [accessibility section](#accessibility).
```

>   Make sure to not use any interactive elements inside the item when using this variant, as it will create nested interactive elements which is not valid HTML and can cause accessibility issues.

Reviewer 2: ```suggestion
  Make sure not to use any interactive elements inside the item when using this variant, as it will create nested interactive elements which is not valid HTML and can cause accessibility issues.
```
